File: packages/r-blockwise/R/brm.R
#' Fit a Blockwise Reduced Modeling (BRM) ensemble
#'
#' BRM partitions the training data into \code{n_blocks} subsets based on
#' per-row patterns of feature-missingness, fits one instance of the supplied
#' \code{learner} per subset using only the features observed in that subset,
#' and at prediction time routes each test row to the subset model whose
#' training-time missingness pattern is closest.
#'
#' The learner interface is intentionally minimal: any \code{fit} /
#' \code{predict} pair can be plugged in via \code{\link{learner}()}, and
#' convenience specs are provided for common families
#' (\code{\link{learner_lm}}, \code{\link{learner_rpart}}, etc.).
#'
#' @param X A data.frame of predictors. May contain \code{NA}. Categorical
#'   predictors should be factors.
#' @param y A numeric vector (regression) or a 0/1 numeric vector
#'   (binary classification) of length \code{nrow(X)}.
#' @param learner A \code{\link{learner}} specification.
#'   Defaults to \code{\link{learner_lm}()}.
#' @param n_blocks Integer number of blocks; if \code{NULL}, chosen
#'   automatically by \code{\link{choose_num_blocks}}.
#' @param low_threshold Column-density threshold for including a predictor in
#'   a block's model. Default 0.05.
#' @param n_restarts k-means restarts for block assignment. Default 5.
#' @param overlap If \code{TRUE} (default) subsets are enlarged via the
#'   set-theoretic inclusion rule; if \code{FALSE}, the non-overlapping
#'   variant is used.
#' @return An object of class \code{"brm"}.
#' @references Srinivasan, K., Currim, F., Ram, S. (2025). A Reduced Modeling
#'   Approach for Making Predictions With Incomplete Data Having Blockwise
#'   Missing Patterns. \emph{INFORMS Journal on Data Science}.
#' @examples
#' \dontrun{
#'   data(bike, package = "blockwise")
#'   bike_miss <- simulate_blockwise_missing(
#'     bike,
#'     blocks       = list(c("hum", "windspeed", "weekday"),
#'                         c("hr", "temp", "weathersit")),
#'     prop_missing = 0.3
#'   )
#'   X <- bike_miss[, setdiff(names(bike_miss), "cnt")]
#'   y <- bike_miss$cnt
#'   fit <- brm(X, y, learner = learner_lm())
#'   preds <- predict(fit, X)
#' }
#' @export
brm <- function(X, y,
                learner = learner_lm(),
                n_blocks = NULL,
                low_threshold = 0.05,
                n_restarts = 5L,
                overlap = TRUE) {
  stopifnot(is.data.frame(X))
  if (nrow(X) != length(y)) {
    stop("X and y must have the same number of observations.")
  }
  if (!inherits(learner, "brm_learner")) {
    stop("`learner` must be a brm_learner object; see ?learner.")
  }

  y_df <- data.frame(y = y)
  names(y_df) <- "__y__"

  if (is.null(n_blocks)) {
    n_blocks <- choose_num_blocks(X, low_threshold = low_threshold)$n_blocks
  }
  n_blocks <- as.integer(n_blocks)

  subs <- build_subsets(X, y_df, n_blocks,
                        low_threshold = low_threshold,
                        n_restarts = n_restarts)
  parts <- if (overlap) subs$overlapping else subs$non_overlapping

  models <- lapply(parts, function(df) {
    X_i <- df[, setdiff(names(df), "__y__"), drop = FALSE]
    y_i <- df[["__y__"]]
    learner$fit(X_i, y_i)
  })

  block_columns <- lapply(subs$columns, function(cc) setdiff(cc, "__y__"))

  out <- list(
    models        = models,
    centers       = subs$centers,
    columns       = block_columns,
    learner       = learner,
    feature_names = names(X),
    n_blocks      = n_blocks,
    overlap       = overlap,
    train_ref     = X
  )
  class(out) <- "brm"
  out
}

#' @export
print.brm <- function(x, ...) {
  cat("Blockwise Reduced Model (BRM)\n")
  cat("  blocks        :", x$n_blocks, "\n")
  cat("  overlap       :", x$overlap, "\n")
  cat("  learner type  :", x$learner$type, "\n")
  cat("  features      :", length(x$feature_names), "\n")
  sz <- vapply(x$columns, length, integer(1L))
  cat("  cols / block  :", paste(sz, collapse = ", "), "\n")
  invisible(x)
}

#' @export
summary.brm <- function(object, ...) {
  structure(
    list(
      n_blocks      = object$n_blocks,
      overlap       = object$overlap,
      learner_type  = object$learner$type,
      feature_names = object$feature_names,
      block_sizes   = vapply(object$columns, length, integer(1L)),
      block_centers = object$centers
    ),
    class = "summary.brm"
  )
}

#' @export
print.summary.brm <- function(x, ...) {
  cat("Blockwise Reduced Model summary\n")
  cat("  n_blocks   :", x$n_blocks, "\n")
  cat("  overlap    :", x$overlap, "\n")
  cat("  learner    :", x$learner_type, "\n")
  cat("  n_features :", length(x$feature_names), "\n\n")
  cat("Feature set per block (1 = observed):\n")
  print(x$block_centers)
  invisible(x)
}

File: packages/py-blockwise/src/blockwise/brm.py
"""The BRM estimator — scikit-learn-compatible."""
from __future__ import annotations

from typing import Optional

import numpy as np
import pandas as pd
from sklearn.base import BaseEstimator, clone
from sklearn.linear_model import LinearRegression

from .blocks import choose_num_blocks
from .impute import hotdeck_impute, impute_with_train
from .subsets import build_subsets


class BRM(BaseEstimator):
    """Blockwise Reduced Modeling estimator.

    BRM partitions the training data into ``n_blocks`` subsets based on
    per-row patterns of feature-missingness, fits one fresh clone of
    ``estimator`` per subset using only the features observed within that
    subset, and at prediction time routes each row of ``X`` to the subset
    model whose training-time missingness center is closest.

    Parameters
    ----------
    estimator : sklearn-style estimator, default=None
        Any object with ``fit(X, y)`` and ``predict(X)``. ``predict_proba(X)``
        is used when available for classification. Cloned per block.
        If None, :class:`sklearn.linear_model.LinearRegression` is used.
    n_blocks : int, optional
        Number of blocks. If ``None``, chosen by :func:`choose_num_blocks`.
    low_threshold : float, default=0.05
        Column-density threshold for including a predictor in a block's model.
    n_restarts : int, default=5
        k-means restarts for block assignment.
    overlap : bool, default=True
        Whether to enlarge each subset using the set-theoretic inclusion
        rule from the paper.

    Attributes
    ----------
    models_ : list
        One fitted estimator per block.
    centers_ : ndarray of shape (n_blocks, n_features)
        Rounded 0/1 missingness centers used to route test rows.
    columns_ : list of list of str
        Per-block feature-column names.
    feature_names_ : list of str
    n_blocks_ : int

    References
    ----------
    Srinivasan, K., Currim, F., Ram, S. (2025). A Reduced Modeling Approach
    for Making Predictions With Incomplete Data Having Blockwise Missing
    Patterns. *INFORMS Journal on Data Science*.
    """

    def __init__(
        self,
        estimator=None,
        n_blocks: Optional[int] = None,
        low_threshold: float = 0.05,
        n_restarts: int = 5,
        overlap: bool = True,
    ):
        self.estimator = estimator
        self.n_blocks = n_blocks
        self.low_threshold = low_threshold
        self.n_restarts = n_restarts
        self.overlap = overlap

    # ---------------------------------------------------------------- fit --
    def fit(self, X, y):
        X = self._as_frame(X)
        y = pd.Series(np.asarray(y), name="__y__")
        if len(X) != len(y):
            raise ValueError("X and y must have the same length.")

        if self.n_blocks is None:
            self.n_blocks_ = choose_num_blocks(
                X, low_threshold=self.low_threshold
            )["n_blocks"]
        else:
            self.n_blocks_ = int(self.n_blocks)

        est = self.estimator if self.estimator is not None else LinearRegression()

        subs = build_subsets(
            X, y, self.n_blocks_,
            low_threshold=self.low_threshold,
            n_restarts=self.n_restarts,
        )
        parts = subs["overlapping"] if self.overlap else subs["non_overlapping"]

        self.models_: list = []
        self.columns_: list[list[str]] = []
        for part in parts:
            cols_x = [c for c in part.columns if c != "__y__"]
            self.columns_.append(cols_x)
            if len(part) == 0 or not cols_x:
                self.models_.append(None)
                continue
            m = clone(est)
            m.fit(part[cols_x], part["__y__"].to_numpy())
            self.models_.append(m)

        self.centers_ = subs["centers"]
        self.feature_names_ = list(X.columns)
        self.train_ref_ = X
        return self

    # ------------------------------------------------------------ predict --
    def predict(self, X) -> np.ndarray:
        X = self._align(X)
        nearest = self._route(X)

        preds = np.empty(len(X), dtype=float)
        preds[:] = np.nan
        for b, model in enumerate(self.models_):
            idx = np.flatnonzero(nearest == b)
            if idx.size == 0 or model is None:
                continue
            cols_b = self.columns_[b]
            X_b = X.iloc[idx][cols_b].copy()
            X_b = impute_with_train(X_b, self.train_ref_[cols_b])
            preds[idx] = np.asarray(model.predict(X_b)).ravel()
        return preds

    def predict_proba(self, X) -> np.ndarray:
        X = self._align(X)
        nearest = self._route(X)

        first = next((m for m in self.models_ if m is not None), None)
        if first is None or not hasattr(first, "predict_proba"):
            raise AttributeError(
                "Underlying estimator does not implement predict_proba."
            )

        # Determine n_classes from the first available model.
        sample_cols = self.columns_[self.models_.index(first)]
        sample_X = (
            self.train_ref_[sample_cols].iloc[:1].pipe(hotdeck_impute)
        )
        n_classes = first.predict_proba(sample_X).shape[1]

        proba = np.zeros((len(X), n_classes))
        for b, model in enumerate(self.models_):
            idx = np.flatnonzero(nearest == b)
            if idx.size == 0 or model is None:
                continue
            cols_b = self.columns_[b]
            X_b = X.iloc[idx][cols_b].copy()
            X_b = impute_with_train(X_b, self.train_ref_[cols_b])
            proba[idx] = model.predict_proba(X_b)
        return proba

    # ------------------------------------------------------------- helpers --
    @staticmethod
    def _as_frame(X) -> pd.DataFrame:
        return X.copy() if isinstance(X, pd.DataFrame) else pd.DataFrame(X)

    def _align(self, X) -> pd.DataFrame:
        X = self._as_frame(X)
        for col in self.feature_names_:
            if col not in X.columns:
                X[col] = np.nan
        return X[self.feature_names_]

    def _route(self, X: pd.DataFrame) -> np.ndarray:
        M = (~X.isna()).astype(np.int8).to_numpy()
        # Squared Euclidean distance to each center.
        dists = ((self.centers_[None, :, :] - M[:, None, :]) ** 2).sum(axis=2)
        return dists.argmin(axis=1)
